File: Haskell/problem4.hs
{- (*) Find the number of elements of a list. -}

import Control.Applicative

len :: [a] -> Int
len = len_ 0

len_ :: Int -> [a] -> Int
len_ acc [] = acc
len_ acc (_:xs) = len_ (acc + 1) xs

len' :: [a] -> Int
len' = foldr (const succ) 0

len'' :: [a] -> Int
len'' xs = sum $ const 1 <$> xs

File: Prolog/Prolog_Lists/problem4.pl
%% 1.04 (*) Find the number of elements of a list.

% len(+Xs, ?Len)
len(Xs, Len) :- len(Xs, 0, Len).

len([], Acc, Acc).
len([_|Xs], Acc, Len) :- NextAcc is Acc + 1, len(Xs, NextAcc, Len).
